ucp/endpoint_reuse.py

```diff
@@ -0,0 +1,125 @@
+import ctypes
+import pickle
+import uuid
+
+from . import core
+
+
+class EPHandle:
+    def __init__(self, ep):
+        self.ep = ep
+        self.refcount = 1
+
+
+class EndpointReuse:
+    """Class to seamlessly reuse endpoints.
+
+    It uses the tag feature of send/recv to separate "virtual" endpoint
+    pairs from each other.
+
+    Warning
+    -------
+    When closing a reused endpoint, the peer might not be notified.
+
+    Performance
+    -----------
+    The overhead of creating endpoints is increased but the performance
+    of the created connections is the same.
+
+    Connection Protocol
+    -------------------
+    1) Client connect to server using a new endpoint.
+    2) Client send the IDs of all its existing endpoints.
+    3) Server receives the IDs from the client and checks if it has a
+       matching existing endpoint. It then sends the matching ID to the
+       client or zero if no match.
+    4) The client and server now continue with either the existing matching
+       endpoints or the new endpoints (which are registered for later reuse).
+    """
+
+    existing_endpoints = {}
+
+    def __init__(self, handle, tag):
+        self.handle = handle
+        self.tag = tag
+
+    @classmethod
+    async def create_endpoint(cls, ip, port):
+        tag = ctypes.c_uint32(uuid.uuid4().int).value
+        ep_new = await core.create_endpoint(ip, port)
+
+        existing_endpoints = list(cls.existing_endpoints.values())
+        my_ep_ids = []
+        for ep in existing_endpoints:
+            if not ep.ep.closed():
+                ep.refcount += 1
+                my_ep_ids.append(ep.ep._msg_tag_recv)
+
+        await ep_new.send_obj(pickle.dumps((my_ep_ids, tag)))
+        reuse_ep_id = pickle.loads(await ep_new.recv_obj())
```

Comment:
Is it possible to get this to work without pickling? Similar question on other lines below.

Reply:
Yes but we have to send a variable length list so I am not sure it is worth the extra complexity? 
Note, this is only part of the connection creation.